libarchive/archive_read_disk_entry_from_file.c

```diff
@@ -419,12 +419,28 @@ setup_acls(struct archive_read_disk *a,
 	if (accpath == NULL)
 		accpath = archive_entry_pathname(entry);
 
+	if (*fd < 0 && a->tree != NULL) {
+		if (a->follow_symlinks ||
+		    archive_entry_filetype(entry) != AE_IFLNK)
+			*fd = a->open_on_current_dir(a->tree,
+			    accpath, O_RDONLY | O_NONBLOCK);
+		if (*fd < 0) {
+			if (a->tree_enter_working_dir(a->tree) != 0) {
+				archive_set_error(&a->archive, errno,
+				    "Couldn't access %s", accpath);
+				return (ARCHIVE_FAILED);
+			}
+		}
+	}
+
```

Comment:
If we don't have a fd and we are inside a subdirectory we need to set `a->tree_enter_working_dir()` otherwise `acl_get_file()` will always lookup a file in the top level directory, which remains current. This code is analogous to the code in `setup_xattrs()`.